FIX several bugs in initial predictions for GBDT

NicolasHug opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Closes #12436 , continuation of the work from @jeremiedbb 

#### What does this implement/fix? Explain your changes.

This PR:

- [X] Fixes the use of the `init` estimator parameter in GBDTs
- [X] exposes `init='zero'` which was supported but undocumented.
- [X] deprecates all the losses and init estimators in `gradient_boosting.py`
  - [X] the loss classes have been moved to a `losses.py` file and have a new `get_init_raw_predictions` method
  - [X] the init estimators are replaced by an equivalent instance of `DummyClassfier` or `DummyRegressor`
- [X] renames `y_pred` into `raw_predictions`, and relevant (private) methods as well.
- [X] fixes some the inaccurate docstrings
- [X] "Fixes" the initial default prediction for multiclass classification: the raw prediction used to be the classes priors, but it should be the log of the priors. In practice the difference in accuracy is very minimal: over 100 runs, I observed an avg improvement in test accuracy of `0.000388` (std = `0.008266`) between the 2 methods, with `n_estimators=1`, `n_samples=10000`.

The `loss.get_init_raw_predictions(X, predictor)` methods return a `raw_prediction` with the correct shape `(n_samples, K)` where `K` is the number of classes in multiclass classification, else `1`.

Those `raw_predictions` are homogeneous to what the trees predict, not homogeneous to what the ensemble predicts. For regression GBDT this is the same thing, but for e.g. binary classification `raw_prediction` is homogeneous to a log-odds ratio.

For more, see https://github.com/scikit-learn/scikit-learn/pull/12436#issuecomment-450037071 and following discussion (not sure if this is much clearer ^^)

#### Any other comments?



<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jeremiedbb committed:
fix grandient boosting with init

jeremiedbb committed:
add solver to avoid deprecation warning

jeremiedbb committed:
optional sample_weight

jeremiedbb committed:
use has_fit_parameter instead

jeremiedbb committed:
typo

jeremiedbb committed:
try except instead of has_fit_parameter

jeremiedbb committed:
fix merge conflicts

jeremiedbb committed:
adress comments

jeremiedbb committed:
support multiclass classification

jeremiedbb committed:
use is_multi_class

jeremiedbb committed:
merge master

NicolasHug committed:
Merge branch 'master' into fix_gbdt_init

NicolasHug committed:
Moved losses into losses.py and deprecated those in gradient_boosting.py

NicolasHug committed:
Done BinomialDeviance

NicolasHug committed:
Done least squares

NicolasHug committed:
Done absolute error

NicolasHug committed:
Done Huber and Quantile losses

NicolasHug committed:
numerical stability for log

NicolasHug committed:
numerical stability for log

NicolasHug committed:
exp loss, pep8

NicolasHug committed:
pep8

NicolasHug committed:
properly exposed the 'zero' init estimator

NicolasHug committed:
Some doc

NicolasHug committed:
Fixed docstrings tests

NicolasHug committed:
Renamed relevant things into raw_predictions

NicolasHug committed:
Updated whatsnew with multiclass classif default init bugfix

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+        Parameters
+        ----------
+        X : array, shape = (n_samples, n_features)
```
```suggestion
        X : 2d array, shape (n_samples, n_features)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+    @abstractmethod
+    def get_init_raw_predictions(self, X, estimator):
+        """
```
You could remove the return line

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+    @abstractmethod
+    def get_init_raw_predictions(self, X, estimator):
+        """
+        Return the initial raw predictions
```
```suggestion
        Return the initial raw predictions.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        X : array, shape = (n_samples, n_features)
+            The data array.
+        estimator : Estimator
+            The estimator to use to compute the predictions
```
```suggestion
            The estimator to use to compute the predictions.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+        Returns
+        -------
+        raw_predictions : 2d array of shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        y : array, shape (n_samples,)
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves).
+
+        sample_weight : array-like, shape (n_samples,), optional
```
is it an `array-like` or just a `1d array`?

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        y : array, shape (n_samples,)
+            The target labels.
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        ----------
+        tree : tree.Tree
+            The tree object.
+        X : array, shape (n, m)
```
```suggestion
        X : 2d array, shape (n, m)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+            The tree object.
+        X : array, shape (n, m)
+            The data array.
+        y : array, shape (n,)
```
```suggestion
        y : 1d array, shape (n,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+            The data array.
+        y : array, shape (n,)
+            The target labels.
+        residual : array, shape (n,)
```
```suggestion
        residual : 1d array, shape (n,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+            The target labels.
+        residual : array, shape (n,)
+            The residuals (usually the negative gradient).
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves) of the
+            tree ensemble at iteration ``i - 1``.
+        sample_weight : array, shape (n,)
```
```suggestion
        sample_weight : 1d array, shape (n,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+            tree ensemble at iteration ``i - 1``.
+        sample_weight : array, shape (n,)
+            The weight of each sample.
+        sample_mask : array, shape (n,)
```
```suggestion
        sample_mask : 1d array, shape (n,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        learning_rate : float, default=0.1
+            learning rate shrinks the contribution of each tree by
+             ``learning_rate``.
+        k : int, default 0
```
```suggestion
        k : int, default=0
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+    @abstractmethod
+    def _update_terminal_region(self, tree, terminal_regions, leaf, X, y,
+                                residual, raw_predictions, sample_weight):
+        """Template method for updating terminal regions (=leaves). """
```
```suggestion
        """Template method for updating terminal regions (i.e., leaves)."""
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+    Parameters
+    ----------
+    n_classes : int
+        Number of classes
```
```suggestion
        Number of classes.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        if not (hasattr(estimator, 'fit') and hasattr(estimator, 'predict')):
+            raise ValueError(
+                "The init parameter must be a valid estimator and "
+                "support both fit and predict"
```
```suggestion
                "support both fit and predict."
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+class LeastSquaresError(RegressionLossFunction):
+    """Loss function for least squares (LS) estimation.
+    Terminal regions need not to be updated for least squares.
```
```suggestion
    Terminal regions do not need to be updated for least squares.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+    Parameters
+    ----------
+    n_classes : int
+        Number of classes
```
```suggestion
        Number of classes.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        Number of classes
+    """
+    def init_estimator(self):
+        # Predicts the median
```
I think that you can remove this comment. Strategy quantile with 0.5 should be self explicit (and this is the Least Absolute Error) as well

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre reviewed (changes requested):
Partial review

glemaitre reviewed (changes requested):
Another partial review

glemaitre reviewed (changes requested):
Another partial review

glemaitre committed:
Update sklearn/ensemble/losses.py

Suggestions from guillaume

Co-Authored-By: NicolasHug <contact@nicolas-hug.com>

glemaitre committed:
Apply suggestions from code review

Co-Authored-By: NicolasHug <contact@nicolas-hug.com>

NicolasHug commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1797,10 +1861,11 @@ class GradientBoostingClassifier(BaseGradientBoosting, ClassifierMixin):
-        If None it uses ``loss.init_estimator``.
+    init : estimator or 'zero', optional (default=None)
+        An estimator object that is used to compute the initial predictions.
+        ``init`` has to provide `fit` and `predict_proba`. If 'zero', the
```
I think we use single backticks for anything that has a glossary entry (that will automatically link it) and double backticks for everything else (to avoid shpinx warnings)

thomasjpfan replied:
@glemaitre Once #13000 gets merged, we do not need to worry about this anymore! Single backticks without a reference will automatically be treated as double backticks!

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,878 @@
+                hasattr(estimator, 'predict_proba')):
+            raise ValueError(
+                "The init parameter must be a valid estimator "
+                "and support both fit and predict_proba".
```
```suggestion
                "and support both fit and predict_proba."
```

glemaitre replied:
Sorry for that one.

NicolasHug commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,875 @@
+        """
+        if sample_weight is None:
+            return np.abs(y - raw_predictions.ravel()).mean()
+        else:
```
Personally I'm not a big fan of this kind of implicit logic.

glemaitre commented:
Regarding the `else`, I don't really mind. Usually `pylint` what encourage for the removal of the `else`.

NicolasHug committed:
Addressed comments from Guillaume

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+    Parameters
+    ----------
+    n_classes : int
+        Number of classes
```
```suggestion
        Number of classes.
```

NicolasHug commented:
Thanks a lot for the review @glemaitre I think I addressed all of the comments

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            True labels
```
```suggestion
            True labels.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        y : array, shape (n_samples,)
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves) of the
```
```suggestion
            The raw predictions (i.e. values from the tree leaves) of the
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves) of the
+            tree ensemble
```
```suggestion
            tree ensemble.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            tree ensemble at iteration ``i - 1``.
+
+        k : int, optional (default=0)
+            The index of the class
```
```suggestion
            The index of the class.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            The raw_predictions (i.e. values from the tree leaves) of the
+            tree ensemble at iteration ``i - 1``.
+
+        k : int, optional (default=0)
```
```suggestion
        k : int, optional default=0
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        y : array, shape (n_samples,)
+            The target labels.
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        numerator *= (self.K - 1) / self.K
+
+        denominator = np.sum(sample_weight * (y - residual) *
+                             (1.0 - y + residual))
```
```suggestion
                             (1 - y + residual))
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            raise ValueError("{0:s} requires 2 classes; got {1:d} class(es)"
+                             .format(self.__class__.__name__, n_classes))
+        # we only need to fit one tree for binary clf.
+        super().__init__(1)
```
```suggestion
        super().__init__(n_classes=1)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            True labels
```
```suggestion
            True labels.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        y : array, shape (n_samples,)
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves) of the
```
```suggestion
            The raw predictions (i.e. values from the tree leaves) of the
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
```
```suggestion
        y : 1d array, shape (n_samples,)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            True labels
```
```suggestion
            True labels.
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        y : array, shape (n_samples,)
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
```
```suggestion
        raw_predictions : 2d array, shape (n_samples, K)
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+            True labels
+
+        raw_predictions : array, shape (n_samples, K)
+            The raw_predictions (i.e. values from the tree leaves) of the
```
```suggestion
            The raw predictions (i.e. values from the tree leaves) of the
```

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        return proba
+
+    def _raw_prediction_to_decision(self, raw_predictions):
+        return (raw_predictions.ravel() >= 0.0).astype(np.int)
```
```suggestion
        return (raw_predictions.ravel() >= 0).astype(np.int)
```

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -29,11 +32,14 @@
 from sklearn.utils.testing import assert_less
 from sklearn.utils.testing import assert_raises
 from sklearn.utils.testing import assert_raise_message
+from sklearn.utils.testing import assert_raises_regex
```
we can use pytest instead

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1324,3 +1319,71 @@ def test_gradient_boosting_validation_fraction():
+
+
+@pytest.mark.parametrize(
+    "task",
```
```suggestion
    "gb, dataset_maker, task",
```

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1324,3 +1319,71 @@ def test_gradient_boosting_validation_fraction():
+     (GradientBoostingClassifier, _make_multiclass, DummyClassifier),
+     (GradientBoostingRegressor, make_regression, DummyRegressor)],
+    ids=["binary classification", "multiclass classification", "regression"])
+def test_gradient_boosting_with_init(task):
```
```suggestion
def test_gradient_boosting_with_init(gb, dataset_maker, task):
```

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1324,3 +1319,71 @@ def test_gradient_boosting_validation_fraction():
+    # estimator.
+    # Check that an error is raised if trying to fit with sample weight but
+    # inital estimator does not support sample weight
+    gb, dataset_maker, init_estimator = task
```
```suggestion
```

glemaitre commented on sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py:
```diff
@@ -100,16 +87,16 @@ def test_sample_weight_init_estimators():
 
         # check if predictions match
-        assert_array_equal(out, sw_out)
+        assert_array_almost_equal(out, sw_out, decimal=2)
```
Numpy is advising to use `assert_allclose`

glemaitre commented on sklearn/utils/testing.py:
```diff
@@ -600,6 +600,8 @@ def uninstall_mldata_mock():
              'CheckingClassifier', 'PatchExtractor', 'CountVectorizer',
              # GradientBoosting base estimators, maybe should
              # exclude them in another way
+             # 0.23: remove all of those that aren't in ensemble.losses
```
prepend a `# FIXME:`

glemaitre commented on doc/whats_new/v0.21.rst:
```diff
@@ -77,10 +77,22 @@ Support for Python 3.4 and below has been officially dropped.
 
+- |Fix| Fixed a bug in :class:`ensemble.GradientBoostingClassifier` and
+  :class:`ensemble.GradientBoostingRegressor`, which didn't support
+  scikit-learn estimators as initial estimator. Also added support of initial
```
```suggestion
  scikit-learn estimators as an initial estimator. Also added support of initial
```

glemaitre reviewed (changes requested):
I would think that parametrizing one of the tests for the error would be nice.
Otherwise it looks good to me.

I'll make a new review once the other changes have been merged.

glemaitre committed:
Apply suggestions from code review

Co-Authored-By: NicolasHug <contact@nicolas-hug.com>

NicolasHug committed:
Addressed comments

glemaitre commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        raw_predictions = raw_predictions.ravel()
+        if sample_weight is None:
+            return -2 * np.mean((y * raw_predictions) -
+                                  np.logaddexp(0, raw_predictions))
```
```suggestion
                                np.logaddexp(0, raw_predictions))
```

NicolasHug committed:
Merge branch 'fix_gbdt_init' of github.com:NicolasHug/scikit-learn into fix_gbdt_init

NicolasHug committed:
pep8

glemaitre reviewed (approved):
LTGM

glemaitre commented:
@jnothman also look at the first issue as well?

jnothman commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+"""Losses and corresponding default initial estimators for gradient boosting
```
Is this move mostly cosmetic? It seems a bit strange to have it sit under `ensemble` without qualifying it as related to GBDT

jnothman replied:
Please confirm that the only changes here are `get_init_raw_predictions` and the init_estimator (I know, I can diff)

NicolasHug replied:
These are the most important changes but also this in losses.py the losses are private, so we can change them how we please.

gradient_boosting.py was is 2500 lines long and I think it really benefits from having the losses in a separate file. Makes reading / editing the code easier.

NicolasHug replied:
(Also I fixed a few docstrings)

jnothman commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+    def _raw_prediction_to_decision(self, raw_predictions):
+        return (raw_predictions.ravel() >= 0).astype(np.int)
+
+    def get_init_raw_predictions(self, X, estimator):
```
If I'm not mistaken this is just `BinomialDeviance.get_init_raw_predictions(self, X, estimator) / 2`... Maybe that should be clearer. Does it make sense to differ from binomial in that way?

jnothman commented on sklearn/ensemble/losses.py:
```diff
@@ -0,0 +1,890 @@
+        val = _weighted_percentile(diff, sample_weight, self.percentile)
+        tree.value[leaf, 0] = val
+
+    def get_init_raw_predictions(self, X, estimator):
```
Can't we just put this defn into RegressionLossFunction?

jnothman reviewed (commented):
starting to look

NicolasHug committed:
put init_raw_pred() in baseclass for regression

jnothman commented:
Is it worth calling the file gb_losses.py?

glemaitre commented:
If the losses should be private, we can recall the file `_gb_losses.py` then?

NicolasHug committed:
Renamed losses.py into _gb_losses.py

glemaitre commented on sklearn/ensemble/_gb_losses.py:
```diff
@@ -0,0 +1,881 @@
+
+        Parameters
+        ----------
+        y : 1d array, shape (n_samples,)
```
@jnothman I went through the discussion in #13000 and other related PR/issues. I am bit behind the topic with my inactivity in the past months. Shall we use the single backtick to refer the glossary documentation. It does not matter here since we are dealing with private classes but for future review it will be useful.

jnothman replied:
I wouldn't force the use of backticks where it's unlikely to help the user. So shape (n_samples,) I'd say we should avoid since it only reduces readability. But in general, yes, we can use backticks to link to glossary when there is no local method by the same name

NicolasHug committed:
Merge branch 'master' into fix_gbdt_init

thomasjpfan commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1396,10 +1440,13 @@ def fit(self, X, y, sample_weight=None, monitor=None):
         X, y = check_X_y(X, y, accept_sparse=['csr', 'csc', 'coo'], dtype=DTYPE)
         n_samples, self.n_features_ = X.shape
+
         if sample_weight is None:
```
Nit: Right above this line, we can set `sample_weight_is_none = sample_weight is None` and use it from then on.

thomasjpfan commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1419,11 +1466,26 @@ def fit(self, X, y, sample_weight=None, monitor=None):
+                if sample_weight_is_none:
+                    self.init_.fit(X, y)
+                else:
+                    if not supports_sample_weight:
```
Nit: We do not need to define `supports_sample_weight` and have it in the if statement here:

```python
if not has_fit_parameter(self.init_, "sample_weight")
```

jeremiedbb replied:
@jnothman argued against using `has_fit_parameter`, see [this comment](https://github.com/scikit-learn/scikit-learn/pull/12436#discussion_r228739257). That's why I used a try except instead:
```python
try:
    self.init_.fit(X, y, sample_weight=sample_weight)
except TypeError:
    if sample_weight_is_none:
        self.init_.fit(X, y)
    else:
        raise ValueError(
            "The initial estimator {} does not support sample "
            "weights.".format(self.init_.__class__.__name__))
```

thomasjpfan reviewed (commented):
This is a good refactor/bug fix. LGTM.

jeremiedbb commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1419,11 +1466,26 @@ def fit(self, X, y, sample_weight=None, monitor=None):
-            self.init_.fit(X, y, sample_weight)
+            # fit initial model and initialize raw predictions
+            if self.init == 'zero':
+                raw_predictions = np.zeros(shape=(X.shape[0], self.loss_.K),
```
should be `self.init_`, right ?

NicolasHug replied:
Both work since we set `self.init_ = self.init` in `_init_state`. I'll use `init_` for consistency

NicolasHug committed:
Merge branch 'master' into fix_gbdt_init

NicolasHug committed:
Addressed Thomas and Jeremie comments

NicolasHug committed:
Merge branch 'master' into fix_gbdt_init

jnothman commented on sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py:
```diff
@@ -100,16 +87,16 @@ def test_sample_weight_init_estimators():
-        assert_array_equal(out, sw_out)
+        assert_allclose(out, sw_out, rtol=1e-2)
 
 
```
Should there be more extensive tests of the new interface to losses?

jnothman reviewed (dismissed):
I can't say I've done my most thorough review ever, but I think this is very good. Not sure if the losses need more tests. Thoughts?

NicolasHug committed:
Added tests for expected output of init_raw_predictions

NicolasHug committed:
Added checks and tests for init estimators output shape

NicolasHug commented:
Thanks @jnothman 

I've added sanity checks for the losses in 2c06512 , and it made me realize some issue that I fixed in e2e1a5c but I need your input (@glemaitre  as well please :) ):

I've added a check in the losses that raises ValueError when the user uses an init estimator that wasn't trained on the same kind of data as the GradientBoosting instance. For example, training the initial estimator on 3 classes while only 2 classes are passed to the GBDT for training. I chose to raise a ValueError with an error message basically saying "This is the user's fault".

The thing is, the ValueError can also be raised because of early stopping: the training data is split into `train` and `val`, and some classes might not be present in `train`: The GBDT loss expects 3 classes but only 2 of them are present in `train`, which the default init estimator will be using. In this case the error message is not clear on what the user should do.

Maybe splitting the data with `train_test_split(stratify=True)` would prevent this?


Note: please ignore the removed #FIXME comments in e2e1a5c, I'll add them back if I can but the fix forces me to remove these checks.

jnothman commented:
stratified splitting could help, but might not if there is a very rare class. Is the problem merely that you don't know how to align the prediction to the validation data? Would initialising the classifier loss with a list of classes known from training facilitate fixing this problem?

NicolasHug commented:
> Is the problem merely that you don't know how to align the prediction to the validation data?

No the problem isn't related to the validation data. It's related to the training data.

-------

Basically the raised ValueErrors in e2e1a5c may happen in 2 scenarios:
- User has passed an init estimator that was trained on a different number of classes than what the GBDT. In this case the error message is appropriate, since this is the user's fault.
- User has passed an init estimator, or is using the default init estimator, AND uses early stopping. If one of the classes is not represented in the actual training data (after `train_test_split`), the init predictor is going to be trained on this data and `init_predictor.predict_proba()` will be of shape (n_samples, n_classes - 1), but the GBDT expects it to be of shape (n_samples, n_classes).

I'm sorry I find it hard to explain clearly and concisely.

I think that using stratify in `train_test_split` would solve the issue because `train_test_split` raises an error when one of the classes are missing in `y_train`. Regardless of this issue, I think that early stopping splits should be stratified anyway. Would that be OK to use it and consider it as a bugfix?

NicolasHug committed:
Should fix FutureWarning errors

jnothman commented:
The init predictor's output can be fixed by aligning it up through classes_
I think. But stratifying does seem a reasonable fix. Does this happen
elsewhere in the library with validation splits?

NicolasHug commented:
>The init predictor's output can be fixed by aligning it up through classes_

Yes but we still need to identify which of the classes are missing in `y_train`, and the loss is not supposed to know anything about `y_pred` at this point (only its init estimator knows about it, but itself cannot know which classes are missing). That's what I mean by tricky / overly complicated.

>Does this happen elsewhere in the library with validation splits?

I don't think the same issues happens in other estimators. This issue is happening because our main estimator (GBDT) relies on another estimator (the init estimator) which is passed only a subset of the data where some classes might be missing, because of early stopping. GBDTs are the only estimators that use early stopping + another estimator, as far as I know.

I took a look at different early stopping implementations:

- All estimators inheriting from BaseSGD (perceptron, passiveagressive, SGDClassifier)  use stratified splits.
- Multilayer perceptron and gradient boosting do not use stratifies splits.

------

Would it be acceptable that I open a new PR to "bugfix" multiplayer perceptron and gradient boosting by making the splits stratified, and wait for it to be merged?

amueller commented:
> Would it be acceptable that I open a new PR to "bugfix" multiplayer perceptron and gradient boosting by making the splits stratified, and wait for it to be merged?

fine with me.

NicolasHug committed:
Handled missing classes after early stopping differently

NicolasHug committed:
Merge branch 'master' into fix_gbdt_init

NicolasHug committed:
EOF

glemaitre commented:
I am still +1 her

jnothman committed:
Merge branch 'master' into fix_gbdt_init

jnothman commented:
Thanks for a phenomenal effort, @NicolasHug!

agramfort commented:
awesome ! thanks heaps @NicolasHug <https://github.com/NicolasHug>!

>
